File: check_all_chats_tasks.py
#!/usr/bin/env python3
"""
Проверка всех чатов на наличие задач и важных обсуждений.

Анализирует последние сообщения из всех чатов и ищет:
- Запросы на действия (@stden сделай, надо, нужно)
- Вопросы требующие ответа
- Упоминания задач, багов, проблем
- Важные обсуждения
"""

import asyncio
import os
from collections import defaultdict
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List

from dotenv import load_dotenv
from telethon import TelegramClient
from telethon.tl.types import Channel

load_dotenv()

API_ID = int(os.getenv("TELEGRAM_API_ID"))
API_HASH = os.getenv("TELEGRAM_API_HASH")
SESSION_FILE = os.getenv("TELEGRAM_SESSION_FILE", "telegram_session")

# Мой username для поиска упоминаний
MY_USERNAME = os.getenv("MY_NAME", "stden")

# Ключевые слова для поиска задач
TASK_KEYWORDS = [
    "сделай",
    "нужно",
    "надо",
    "можешь",
    "помоги",
    "исправь",
    "добавь",
    "удали",
    "настрой",
    "проверь",
    "запусти",
    "@" + MY_USERNAME,
]

# Ключевые слова для вопросов
QUESTION_KEYWORDS = ["?", "как", "почему", "что делать", "где", "когда", "кто"]

# Ключевые слова для проблем
PROBLEM_KEYWORDS = ["не работает", "ошибка", "баг", "сломалось", "проблема", "упало", "крашится", "зависло"]


class ChatTaskChecker:
    """Проверка задач во всех чатах."""

    def __init__(self):
        self.client: TelegramClient = None
        self.tasks_by_chat: Dict[str, List[Dict]] = defaultdict(list)
        self.questions_by_chat: Dict[str, List[Dict]] = defaultdict(list)
        self.problems_by_chat: Dict[str, List[Dict]] = defaultdict(list)

    async def __aenter__(self):
        """Connect to Telegram."""
        self.client = TelegramClient(SESSION_FILE, API_ID, API_HASH)
        await self.client.connect()

        if not await self.client.is_user_authorized():
            raise RuntimeError("Not authorized. Run `cargo run -- init-session` first.")

        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Disconnect from Telegram."""
        if self.client:
            await self.client.disconnect()

    async def check_all_chats(self, days_back: int = 3, limit_per_chat: int = 100):
        """Check all chats for tasks.

        Args:
            days_back: How many days back to check
            limit_per_chat: Maximum messages per chat
        """
        print(f"\n🔍 Проверяю чаты за последние {days_back} дней...")
        print(f"Лимит сообщений на чат: {limit_per_chat}\n")

        offset_date = datetime.now() - timedelta(days=days_back)

        # Получаем все диалоги
        dialogs = await self.client.get_dialogs()

        # Фильтруем только каналы и группы
        relevant_dialogs = [d for d in dialogs if isinstance(d.entity, Channel) and not d.entity.broadcast]

        print(f"Найдено релевантных чатов: {len(relevant_dialogs)}\n")

        for dialog in relevant_dialogs:
            try:
                chat_name = dialog.title or "Unknown"
                await self.check_chat(dialog.entity, chat_name, offset_date, limit_per_chat)
            except Exception as e:
                print(f"❌ Ошибка при проверке '{dialog.title}': {e}")

    async def check_chat(self, entity, chat_name: str, offset_date: datetime, limit: int):
        """Check single chat for tasks.

        Args:
            entity: Telegram entity
            chat_name: Chat name
            offset_date: Date to start from
            limit: Message limit
        """
        message_count = 0
        tasks_found = 0
        questions_found = 0
        problems_found = 0

        async for message in self.client.iter_messages(entity, limit=limit, offset_date=offset_date):
            if not message.message:
                continue

            message_count += 1
            text = message.message.lower()

            # Check for tasks
            if self._is_task(text, message.message):
                self.tasks_by_chat[chat_name].append(
                    {
                        "text": message.message,
                        "date": message.date,
                        "sender": await self._get_sender_name(message),
                        "id": message.id,
                    }
                )
                tasks_found += 1

            # Check for questions
            if self._is_question(text):
                self.questions_by_chat[chat_name].append(
                    {
                        "text": message.message,
                        "date": message.date,
                        "sender": await self._get_sender_name(message),
                        "id": message.id,
                    }
                )
                questions_found += 1

            # Check for problems
            if self._is_problem(text):
                self.problems_by_chat[chat_name].append(
                    {
                        "text": message.message,
                        "date": message.date,
                        "sender": await self._get_sender_name(message),
                        "id": message.id,
                    }
                )
                problems_found += 1

        if tasks_found > 0 or questions_found > 0 or problems_found > 0:
            print(f"📌 {chat_name}")
            if tasks_found > 0:
                print(f"   ✅ Задач: {tasks_found}")
            if questions_found > 0:
                print(f"   ❓ Вопросов: {questions_found}")
            if problems_found > 0:
                print(f"   ⚠️  Проблем: {problems_found}")
            print(f"   📊 Проверено сообщений: {message_count}\n")

    def _is_task(self, text_lower: str, original_text: str) -> bool:
        """Check if message contains a task."""
        # Check for mentions
        if f"@{MY_USERNAME.lower()}" in text_lower:
            return True

        # Check for task keywords
        for keyword in TASK_KEYWORDS:
            if keyword.lower() in text_lower and len(original_text) > 20:
                return True

        return False

    def _is_question(self, text: str) -> bool:
        """Check if message is a question."""
        for keyword in QUESTION_KEYWORDS:
            if keyword in text:
                return True
        return False

    def _is_problem(self, text: str) -> bool:
        """Check if message describes a problem."""
        for keyword in PROBLEM_KEYWORDS:
            if keyword in text:
                return True
        return False

    async def _get_sender_name(self, message) -> str:
        """Get sender name from message."""
        if not message.sender:
            return "Unknown"

        sender = message.sender
        if hasattr(sender, "first_name"):
            return f"{sender.first_name or ''} {sender.last_name or ''}".strip() or "Unknown"
        elif hasattr(sender, "title"):
            return sender.title
        return "Unknown"

    def generate_report(self, output_file: Path):
        """Generate markdown report.

        Args:
            output_file: Path to save report
        """
        lines = []

        # Header
        lines.append("# 📋 Задачи и обсуждения из всех чатов")
        lines.append("")
        lines.append(f"**Дата проверки:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        lines.append("")

        # Summary
        total_tasks = sum(len(tasks) for tasks in self.tasks_by_chat.values())
        total_questions = sum(len(q) for q in self.questions_by_chat.values())
        total_problems = sum(len(p) for p in self.problems_by_chat.values())

        lines.append("## 📊 Общая статистика")
        lines.append("")
        lines.append(
            f"- **Чатов проверено:** {len(set(list(self.tasks_by_chat.keys()) + list(self.questions_by_chat.keys()) + list(self.problems_by_chat.keys())))}"
        )
        lines.append(f"- **Задач найдено:** {total_tasks}")
        lines.append(f"- **Вопросов найдено:** {total_questions}")
        lines.append(f"- **Проблем найдено:** {total_problems}")
        lines.append("")

        # Tasks by chat
        if self.tasks_by_chat:
            lines.append("## ✅ Задачи по чатам")
            lines.append("")

            for chat_name, tasks in sorted(self.tasks_by_chat.items(), key=lambda x: len(x[1]), reverse=True):
                lines.append(f"### {chat_name} ({len(tasks)} задач)")
                lines.append("")

                for task in sorted(tasks, key=lambda x: x["date"], reverse=True)[:5]:
                    date_str = task["date"].strftime("%Y-%m-%d %H:%M")
                    lines.append(f"**[{date_str}] {task['sender']}:**")
                    lines.append(f"> {task['text'][:200]}{'...' if len(task['text']) > 200 else ''}")
                    lines.append("")

        # Questions by chat
        if self.questions_by_chat:
            lines.append("## ❓ Вопросы по чатам")
            lines.append("")

            for chat_name, questions in sorted(self.questions_by_chat.items(), key=lambda x: len(x[1]), reverse=True):
                if len(questions) < 3:  # Skip chats with few questions
                    continue

                lines.append(f"### {chat_name} ({len(questions)} вопросов)")
                lines.append("")

                for q in sorted(questions, key=lambda x: x["date"], reverse=True)[:3]:
                    date_str = q["date"].strftime("%Y-%m-%d %H:%M")
                    lines.append(f"**[{date_str}] {q['sender']}:**")
                    lines.append(f"> {q['text'][:200]}{'...' if len(q['text']) > 200 else ''}")
                    lines.append("")

        # Problems by chat
        if self.problems_by_chat:
            lines.append("## ⚠️ Проблемы по чатам")
            lines.append("")

            for chat_name, problems in sorted(self.problems_by_chat.items(), key=lambda x: len(x[1]), reverse=True):
                lines.append(f"### {chat_name} ({len(problems)} проблем)")
                lines.append("")

                for p in sorted(problems, key=lambda x: x["date"], reverse=True)[:5]:
                    date_str = p["date"].strftime("%Y-%m-%d %H:%M")
                    lines.append(f"**[{date_str}] {p['sender']}:**")
                    lines.append(f"> {p['text'][:200]}{'...' if len(p['text']) > 200 else ''}")
                    lines.append("")

        # Save
        output_file.parent.mkdir(parents=True, exist_ok=True)
        output_file.write_text("\n".join(lines), encoding="utf-8")

        print(f"\n✅ Отчёт сохранён: {output_file}")


async def main():
    """Main function."""
    import argparse

    parser = argparse.ArgumentParser(description="Проверить все чаты на задачи")
    parser.add_argument("--days", type=int, default=3, help="Сколько дней назад проверять (по умолчанию: 3)")
    parser.add_argument("--limit", type=int, default=100, help="Максимум сообщений на чат (по умолчанию: 100)")
    parser.add_argument(
        "--output", type=Path, default=Path("analysis_results/all_chats_tasks.md"), help="Путь для сохранения отчёта"
    )

    args = parser.parse_args()

    async with ChatTaskChecker() as checker:
        await checker.check_all_chats(args.days, args.limit)
        checker.generate_report(args.output)

    print("\n✨ Готово!")


if __name__ == "__main__":
    asyncio.run(main())

File: src/bin/analysis/check_all_chats_tasks.rs
//! Check all chats for tasks and important discussions.
//!
//! Analyzes recent messages from all chats and searches for:
//! - Action requests (@your_username do, need, must)
//! - Questions requiring answers
//! - Mentions of tasks, bugs, problems
//! - Important discussions
//!
//! Usage: cargo run --bin check_all_chats_tasks -- [--days 3] [--limit 100] [--output analysis_results/all_chats_tasks.md]

use anyhow::Result;
use chrono::{DateTime, Duration, Utc};
use clap::Parser;
use dotenvy::dotenv;
use grammers_client::types::{Message, Peer};
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::PathBuf;
use telegram_reader::chat::peer_name;
use telegram_reader::{get_client, SessionLock};

#[derive(Parser)]
struct Args {
    /// How many days back to check
    #[arg(long, default_value = "3")]
    days: i64,

    /// Maximum messages per chat
    #[arg(long, default_value = "100")]
    limit: usize,

    /// Path for saving the report
    #[arg(long, default_value = "analysis_results/all_chats_tasks.md")]
    output: PathBuf,
}

#[derive(Debug, Clone)]
struct Entry {
    text: String,
    date: DateTime<Utc>,
    sender: String,
}

struct ChatTaskChecker {
    my_username: String,
    tasks_by_chat: HashMap<String, Vec<Entry>>,
    questions_by_chat: HashMap<String, Vec<Entry>>,
    problems_by_chat: HashMap<String, Vec<Entry>>,
}

const TASK_KEYWORDS: &[&str] = &[
    "сделай",
    "нужно",
    "надо",
    "можешь",
    "помоги",
    "исправь",
    "добавь",
    "удали",
    "настрой",
    "проверь",
    "запусти",
];

const QUESTION_KEYWORDS: &[&str] = &["?", "как", "почему", "что делать", "где", "когда", "кто"];

const PROBLEM_KEYWORDS: &[&str] = &[
    "не работает",
    "ошибка",
    "баг",
    "сломалось",
    "проблема",
    "упало",
    "крашится",
    "зависло",
];

impl ChatTaskChecker {
    fn new(my_username: String) -> Self {
        Self {
            my_username,
            tasks_by_chat: HashMap::new(),
            questions_by_chat: HashMap::new(),
            problems_by_chat: HashMap::new(),
        }
    }

    async fn check_chat(
        &mut self,
        _client: &grammers_client::Client,
        message: &Message,
        chat_name: &str,
    ) -> Result<()> {
        let text = message.text();
        if text.is_empty() {
            return Ok(());
        }

        let text_lower = text.to_lowercase();
        let sender = if let Some(s) = message.sender() {
            peer_name(s)
        } else {
            "Unknown".to_string()
        };

        let entry = Entry {
            text: text.to_string(),
            date: message.date(),
            sender,
        };

        // Check for tasks
        if text_lower.contains(&format!("@{}", self.my_username.to_lowercase()))
            || (TASK_KEYWORDS.iter().any(|k| text_lower.contains(k)) && text.len() > 20)
        {
            self.tasks_by_chat
                .entry(chat_name.to_string())
                .or_default()
                .push(entry.clone());
        }

        // Check for questions
        if QUESTION_KEYWORDS.iter().any(|k| text_lower.contains(k)) {
            self.questions_by_chat
                .entry(chat_name.to_string())
                .or_default()
                .push(entry.clone());
        }

        // Check for problems
        if PROBLEM_KEYWORDS.iter().any(|k| text_lower.contains(k)) {
            self.problems_by_chat
                .entry(chat_name.to_string())
                .or_default()
                .push(entry);
        }

        Ok(())
    }

    fn generate_report(&self, output_file: &PathBuf) -> Result<()> {
        let mut lines = Vec::new();

        lines.push("# 📋 Задачи и обсуждения из всех чатов".to_string());
        lines.push("".to_string());
        lines.push(format!(
            "**Дата проверки:** {}",
            Utc::now().format("%Y-%m-%d %H:%M:%S")
        ));
        lines.push("".to_string());

        let all_chats: std::collections::HashSet<_> = self
            .tasks_by_chat
            .keys()
            .chain(self.questions_by_chat.keys())
            .chain(self.problems_by_chat.keys())
            .collect();

        let total_tasks: usize = self.tasks_by_chat.values().map(|v| v.len()).sum();
        let total_questions: usize = self.questions_by_chat.values().map(|v| v.len()).sum();
        let total_problems: usize = self.problems_by_chat.values().map(|v| v.len()).sum();

        lines.push("## 📊 Общая статистика".to_string());
        lines.push("".to_string());
        lines.push(format!("- **Чатов проверено:** {}", all_chats.len()));
        lines.push(format!("- **Задач найдено:** {}", total_tasks));
        lines.push(format!("- **Вопросов найдено:** {}", total_questions));
        lines.push(format!("- **Проблем найдено:** {}", total_problems));
        lines.push("".to_string());

        if !self.tasks_by_chat.is_empty() {
            lines.push("## ✅ Задачи по чатам".to_string());
            lines.push("".to_string());

            let mut sorted_chats: Vec<_> = self.tasks_by_chat.iter().collect();
            sorted_chats.sort_by_key(|x| std::cmp::Reverse(x.1.len()));

            for (chat_name, tasks) in sorted_chats {
                lines.push(format!("### {} ({} задач)", chat_name, tasks.len()));
                lines.push("".to_string());

                let mut sorted_tasks = tasks.clone();
                sorted_tasks.sort_by_key(|x| std::cmp::Reverse(x.date));

                for task in sorted_tasks.iter().take(5) {
                    lines.push(format!(
                        "**[{}] {}:**",
                        task.date.format("%Y-%m-%d %H:%M"),
                        task.sender
                    ));
                    let preview = if task.text.len() > 200 {
                        format!("{}...", &task.text[..200])
                    } else {
                        task.text.clone()
                    };
                    lines.push(format!("> {}", preview.replace('\n', "\n> ")));
                    lines.push("".to_string());
                }
            }
        }

        if !self.questions_by_chat.is_empty() {
            lines.push("## ❓ Вопросы по чатам".to_string());
            lines.push("".to_string());

            let mut sorted_chats: Vec<_> = self.questions_by_chat.iter().collect();
            sorted_chats.sort_by_key(|x| std::cmp::Reverse(x.1.len()));

            for (chat_name, questions) in sorted_chats {
                if questions.len() < 3 {
                    continue;
                }
                lines.push(format!("### {} ({} вопросов)", chat_name, questions.len()));
                lines.push("".to_string());

                let mut sorted_q = questions.clone();
                sorted_q.sort_by_key(|x| std::cmp::Reverse(x.date));

                for q in sorted_q.iter().take(3) {
                    lines.push(format!(
                        "**[{}] {}:**",
                        q.date.format("%Y-%m-%d %H:%M"),
                        q.sender
                    ));
                    let preview = if q.text.len() > 200 {
                        format!("{}...", &q.text[..200])
                    } else {
                        q.text.clone()
                    };
                    lines.push(format!("> {}", preview.replace('\n', "\n> ")));
                    lines.push("".to_string());
                }
            }
        }

        if !self.problems_by_chat.is_empty() {
            lines.push("## ⚠️ Проблемы по чатам".to_string());
            lines.push("".to_string());

            let mut sorted_chats: Vec<_> = self.problems_by_chat.iter().collect();
            sorted_chats.sort_by_key(|x| std::cmp::Reverse(x.1.len()));

            for (chat_name, problems) in sorted_chats {
                lines.push(format!("### {} ({} проблем)", chat_name, problems.len()));
                lines.push("".to_string());

                let mut sorted_p = problems.clone();
                sorted_p.sort_by_key(|x| std::cmp::Reverse(x.date));

                for p in sorted_p.iter().take(5) {
                    lines.push(format!(
                        "**[{}] {}:**",
                        p.date.format("%Y-%m-%d %H:%M"),
                        p.sender
                    ));
                    let preview = if p.text.len() > 200 {
                        format!("{}...", &p.text[..200])
                    } else {
                        p.text.clone()
                    };
                    lines.push(format!("> {}", preview.replace('\n', "\n> ")));
                    lines.push("".to_string());
                }
            }
        }

        if let Some(parent) = output_file.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::write(output_file, lines.join("\n"))?;

        println!("\n✅ Отчёт сохранён: {:?}", output_file);
        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    dotenv().ok();
    tracing_subscriber::fmt::init();

    let args = Args::parse();
    let my_username = env::var("MY_NAME").unwrap_or_else(|_| "your_username".to_string());

    let _lock = SessionLock::acquire()?;
    let client = get_client().await?;

    println!("\n🔍 Проверяю чаты за последние {} дней...", args.days);
    println!("Лимит сообщений на чат: {}\n", args.limit);

    let offset_date = Utc::now() - Duration::days(args.days);
    let mut checker = ChatTaskChecker::new(my_username);

    let mut dialogs = client.iter_dialogs();
    while let Some(dialog) = dialogs.next().await? {
        // Only channels and groups
        if matches!(dialog.peer, Peer::User(_)) {
            continue;
        }
        // Skip broadcast channels
        if let Peer::Channel(ref c) = dialog.peer {
            if !c.raw.megagroup && c.raw.broadcast {
                continue;
            }
        }

        let chat_name = peer_name(&dialog.peer);
        let mut messages = client.iter_messages(&dialog.peer).limit(args.limit);
        let mut count = 0;
        let mut found_anything = false;

        while let Some(message) = messages.next().await? {
            if message.date() < offset_date {
                break;
            }
            checker.check_chat(&client, &message, &chat_name).await?;
            count += 1;
            found_anything = true;
        }

        if found_anything {
            println!("📌 {} (проверено {})", chat_name, count);
        }
    }

    checker.generate_report(&args.output)?;

    println!("\n✨ Готово!");
    Ok(())
}
